docs/smart_bits.md

```diff
@@ -0,0 +1,71 @@
+# SmartBits
+
+## Overview
+"Smart" types in magma (`SmartBits`, `SmartBit`) emulate verilog bits types, allowing very compact syntax for describing datapaths. For example, we can simply write the following:
+
+```python
+x = m.smart.SmartBits[7]()
+y = m.smart.SmartBits[9]()
+z = m.smart.SmartBits[12]()
+
+expr = (~(x + y) + z) << x
+
+a @= expr
+```
+
+Note that using raw magma types (i.e. `Bits`, `UInt`, `SInt`), we would have to do several explicit extensions and explicit promotion to `UInt` (for example, because addition is not defined on plain `Bits`). The smart types allow for writing this simple syntax while still compiling to raw magma types (and therefore the generated code still adheres to the typing rules of the raw magma types).
+
+It should also be noted that due to the implementation details, expressions on smart types are not actually operable magma types. In fact, the only thing you can do with them is wire them to other smart types, as in `a @= expr`. At the time of wiring, bit-widths and signedness of operands are resolved as a function of the expression itself and the left hand side of the assignment (see below for details). So, a statement like `b @= expr`, means that the generated RTL could different based on the bit-width of `b`.
+
+## Types
+There are two smart type constructors:
+* `SmartBits[bit_width, signed=False]`: returns a SmartBits type with bit-width `bit_width` and signedness `sign`. `sign` defaults to `False`.
+* `SmartBit` which is a literal alias of `SmartBits[1]`.
+
+Expressions built on top of these base smart types are of an intermediate type `SmartExpr`, which can only (a) be used in another expression, such as `dbl = expr + expr`, or (b) be on the right hand side of a wiring assignment. Besides these two operations (and calling `str()` for debugging), these intermediate objects are not meant to be introspected or operated on by the user.
+
+## Operators
+Arithmetic, bit-wise logical, comparison, shift, bit-wise reduction, and concat operators are defined for all smart types and intermediates. All operators are implemented as python-style infix operators (e.g. `a + b`), except for the following:
+* Concat: the function `concat` is a free, variadic function; e.g. `long = concat(x, y, z)`
+* Bit-wise reduce: bit-wise reductions are defined as a member function `reduce`, which takes the reduction operator as an argument; e.g. `short = x.reduce(operator.or_)`
```

Comment:
We could add `m.reduce()` support for this pattern, but this should work for now

Reply:
Good point. Wanted to make this code path decoupled from rest of existing magma. Once this is more mainline, there should be more integrated first class support, such as in `m.reduce()`.